runtimes/kotlin-1.6/src/main/kotlin/io/openruntimes/kotlin/Server.kt

```diff
@@ -8,56 +8,173 @@ import java.io.StringWriter
+
+    for(param in queryString.split("&")) {
+        var pair = param.split("=", limit = 2)
+
+        if(pair.size >= 1 && !pair[0].isEmpty()) {
+            var value: String = if (pair.size == 2) pair[1] else ""
+            query.put(pair[0], value)
+        }
+    }
+
+    var url: String = scheme + "://" + host;
+
+    if(port != Integer.parseInt(defaultPort)) {
+        url += ":" + port;
+    }
+
+    url += path;
 
-    val outStream = ByteArrayOutputStream()
-    val errStream = ByteArrayOutputStream()
-    val userOut = PrintStream(outStream)
-    val userErr = PrintStream(errStream)
-    val systemOut = System.out
-    val systemErr = System.err
+    if(!queryString.isEmpty()) {
+        url += "?" + queryString;
+    }
+
+    var runtimeRequest: RuntimeRequest = RuntimeRequest(bodyString, body, headers, method, url, path, host, scheme, port, queryString, query)
+    var runtimeResponse: RuntimeResponse = RuntimeResponse()
+    var context: RuntimeContext = RuntimeContext(runtimeRequest, runtimeResponse)
+
+    var systemOut = System.out;
+    var systemErr = System.err;
 
-    System.setOut(userOut)
-    System.setErr(userErr)
+    var customstdStream: ByteArrayOutputStream = ByteArrayOutputStream();
+    var customstd: PrintStream = PrintStream(customstdStream);
+    System.setOut(customstd);
+    System.setErr(customstd);
+
+    var output: RuntimeOutput?
 
     try {
-        val userResponse = codeWrapper.main(request, response)
-        val output = mutableMapOf(
-            "response" to userResponse.data,
-            "stdout" to outStream.toString()
-        )
-        ctx.json(output)
+        var classToLoad = Class.forName("io.openruntimes.kotlin.Tests");
+        var classMethod = classToLoad.getDeclaredMethod("main", RuntimeContext::class.java);
+        var instance = classToLoad.newInstance();
+        output = classMethod.invoke(instance, context) as RuntimeOutput;
     } catch (e: Exception) {
-        e.printStackTrace()
-        val output = mutableMapOf(
-            "stdout" to outStream.toString(),
-            "stderr" to errStream.toString()
-        )
-        ctx.status(500).json(output)
+        var sw: StringWriter = StringWriter();
+        var pw: PrintWriter = PrintWriter(sw);
+        e.printStackTrace(pw);
+
+        context.error(sw.toString());
+        output = context.res.send("", 500, mutableMapOf<String, String>());
     } finally {
-         System.out.flush();
-         System.err.flush();
-         System.setOut(systemOut);
-         System.setErr(systemErr);
+        System.out.flush();
+        System.err.flush();
+        System.setOut(systemOut);
+        System.setErr(systemErr);
     }
+
+    if(output == null) {
+        context.error("Return statement missing. return context.res.empty() if no response is expected.");
+        output = context.res.send("", 500, mutableMapOf<String, String>());
+    }
+
+    for (entry in output.headers.entries.iterator()) {
```

Comment:
You can use destructuring for maps in for loops in Kotlin:

```suggestion
    for ((header, value) in output.headers) {
```